File: backend/src/schemas.py
from __future__ import annotations

from datetime import timedelta
from uuid import UUID

from pydantic import BaseModel

from src.models.groups import Area, Group, Meet
from src.models.notifications import Notification
from src.models.users import FileModel, Gender, User


class UserBaseSchema(BaseModel):
    user_id: str
    name: str
    email: str
    phone: str
    gender: Gender
    date_of_birth: str
    description: str

    @staticmethod
    def from_model(user: User) -> UserBaseSchema:
        return UserBaseSchema(
            user_id=str(user.user_id),
            name=user.name,
            email=user.email,
            phone=user.phone,
            gender=user.gender,
            date_of_birth=user.date_of_birth,
            description=user.description,
        )


class UserSchema(BaseModel):
    """User schema for the API"""

    user_id: str
    name: str
    email: str
    phone: str
    gender: Gender
    date_of_birth: str
    description: str
    is_coach: bool

    @staticmethod
    def from_model(user: User) -> UserSchema:
        return UserSchema(
            user_id=str(user.user_id),
            name=user.name,
            email=user.email,
            phone=user.phone,
            gender=user.gender,
            date_of_birth=user.date_of_birth,
            description=user.description,
            is_coach=user.is_coach,
        )


class FileSchema(BaseModel):
    file_id: str
    name: str

    @staticmethod
    def from_model(file: FileModel) -> FileSchema:
        return FileSchema(
            file_id=str(file.file_id),
            name=file.name,
        )


class CertificatesSchema(BaseModel):
    certificates: list[FileSchema]

    @staticmethod
    def from_model(certificates: list[FileModel]) -> CertificatesSchema:
        return CertificatesSchema(
            certificates=[FileSchema.from_model(certificate) for certificate in certificates],
        )


class AreaSchema(BaseModel):
    area_id: str
    name: str

    @staticmethod
    def from_model(area: Area) -> AreaSchema:
        return AreaSchema(
            area_id=str(area.area_id),
            name=area.name,
        )


class LoginResponseSchema(BaseModel):
    auth_token: UUID
    user: UserSchema
    areas: list[AreaSchema]


class GroupSchema(BaseModel):
    group_id: str
    coach_id: str
    coach_name: str
    name: str
    description: str
    area_id: str

    @staticmethod
    def from_model(group: Group, coach_name: str) -> GroupSchema:
        return GroupSchema(
            group_id=str(group.group_id),
            coach_id=str(group.coach_id),
            coach_name=coach_name,
            name=group.name,
            description=group.description,
            area_id=str(group.area_id),
        )


class GroupInfoSchema(BaseModel):
    group_id: str
    coach_name: str
    name: str
    area_name: str

    @staticmethod
    def from_model(row: tuple[UUID, str, str, str]) -> GroupInfoSchema:
        return GroupInfoSchema(
            group_id=str(row[0]),
            coach_name=row[1],
            name=row[2],
            area_name=row[3],
        )


class MeetSchema(BaseModel):
    meet_id: str
    group_id: str
    group_name: str
    max_members: int
    meet_date: str
    start_time: str
    end_time: str
    duration: int
    city: str
    street: str

    members: list[UserBaseSchema]

    @staticmethod
    def from_model(meet: Meet, group_name: str, members: list[User]) -> MeetSchema:
        end_time = meet.meet_date + timedelta(minutes=meet.duration)

        return MeetSchema(
            meet_id=str(meet.meet_id),
            group_id=str(meet.group_id),
            group_name=group_name,
            max_members=meet.max_members,
            meet_date=meet.meet_date.date().strftime("%Y-%m-%d %H:%M:%S"),
            start_time=meet.meet_date.time().strftime("%Y-%m-%d %H:%M:%S"),
            end_time=end_time.time().strftime("%Y-%m-%d %H:%M:%S"),
            duration=meet.duration,
            city=meet.city,
            street=meet.street,
            members=[UserBaseSchema.from_model(member) for member in members],
        )


class MeetInfoSchema(BaseModel):
    meet_id: str
    group_id: str
    group_name: str
    meet_date: str
    start_time: str
    end_time: str
    duration: int
    city: str
    street: str
    full: bool
    registered: bool

    @staticmethod
    def from_model(meet: Meet, group_name: str, full: bool, registered: bool) -> MeetInfoSchema:
        end_time = meet.meet_date + timedelta(minutes=meet.duration)

        return MeetInfoSchema(
            meet_id=str(meet.meet_id),
            group_id=str(meet.group_id),
            group_name=group_name,
            meet_date=meet.meet_date.date().strftime("%Y-%m-%d %H:%M:%S"),
            start_time=meet.meet_date.time().strftime("%Y-%m-%d %H:%M:%S"),
            end_time=end_time.time().strftime("%Y-%m-%d %H:%M:%S"),
            duration=meet.duration,
            city=meet.city,
            street=meet.street,
            full=full,
            registered=registered,
        )


class GroupViewInfoSchema(BaseModel):
    group: GroupSchema
    meets: list[MeetInfoSchema]
    registered: bool


class GroupFullSchema(BaseModel):
    group: GroupSchema
    meets: list[MeetSchema]
    members: list[UserBaseSchema]

    @staticmethod
    def from_model(group: Group, coach_name: str, meets: list[Meet], members: list[User]) -> GroupFullSchema:
        return GroupFullSchema(
            group=GroupSchema.from_model(group, coach_name),
            meets=[MeetSchema.from_model(meet, group.name, []) for meet in meets],
            members=[UserBaseSchema.from_model(member) for member in members],
        )


class MeetViewInfoSchema(BaseModel):
    group: GroupSchema
    meet: MeetInfoSchema


class MyGroupsSchema(BaseModel):
    in_groups: list[GroupInfoSchema]
    coach_groups: list[GroupSchema]


class MyMeetsSchema(BaseModel):
    meets: list[MeetInfoSchema]


class ViewCoachSchema(BaseModel):
    coach: UserBaseSchema
    certificates: list[FileSchema]

    @staticmethod
    def from_model(coach: User, certificates: list[FileModel]) -> ViewCoachSchema:
        return ViewCoachSchema(
            coach=UserBaseSchema.from_model(coach),
            certificates=[FileSchema.from_model(certificate) for certificate in certificates],
        )


class NotificationSchema(BaseModel):
    notification_id: str
    message: str
    date: str

    @staticmethod
    def from_model(notification: Notification) -> NotificationSchema:
        return NotificationSchema(
            notification_id=str(notification.notification_id),
            message=notification.message,
            date=notification.date.strftime("%Y-%m-%d %H:%M:%S"),
        )


class NotificationsSchema(BaseModel):
    notifications: list[NotificationSchema]

    @staticmethod
    def from_model(notifications: list[Notification]) -> NotificationsSchema:
        return NotificationsSchema(
            notifications=[NotificationSchema.from_model(notification) for notification in notifications],
        )

File: mobile/lib/schemas.dart
// This file contains the API Schemas

class UserBaseSchema {
  final String userId;
  final String name;
  final String email;
  final String phone;
  final String gender;
  final DateTime dateOfBirth;

  UserBaseSchema(this.userId, this.name, this.email, this.phone, this.gender,
      this.dateOfBirth);

  factory UserBaseSchema.fromJson(dynamic data) {
    return UserBaseSchema(
      data['user_id'] as String,
      data['name'] as String,
      data['email'] as String,
      data['phone'] as String,
      data['gender'] as String,
      DateTime.parse(data['date_of_birth'] as String),
    );
  }
}

class UserSchema {
  final String userId;
  final String name;
  final String email;
  final String phone;
  final String gender;
  final DateTime dateOfBirth;
  final String description;
  final bool isCoach;

  UserSchema(this.userId, this.name, this.email, this.phone, this.gender,
      this.dateOfBirth, this.description, this.isCoach);

  factory UserSchema.fromJson(dynamic data) {
    return UserSchema(
      data['user_id'] as String,
      data['name'] as String,
      data['email'] as String,
      data['phone'] as String,
      data['gender'] as String,
      DateTime.parse(data['date_of_birth'] as String),
      data['description'] as String,
      data['is_coach'] as bool,
    );
  }
}

class FileSchema {
  final String fileId;
  final String name;

  FileSchema(this.fileId, this.name);

  factory FileSchema.fromJson(dynamic data) {
    return FileSchema(
      data['file_id'] as String,
      data['name'] as String,
    );
  }
}

class CertificatesSchema {
  final List<FileSchema> certificates;

  CertificatesSchema(this.certificates);

  factory CertificatesSchema.fromJson(dynamic data) {
    return CertificatesSchema(
      (data['certificates'] as List<dynamic>)
          .map((file) => FileSchema.fromJson(file))
          .toList(),
    );
  }
}

class AreaSchema {
  final String areaId;
  final String name;

  AreaSchema(this.areaId, this.name);

  factory AreaSchema.fromJson(dynamic data) {
    return AreaSchema(
      data['area_id'] as String,
      data['name'] as String,
    );
  }
}

class LoginResponseSchema {
  final String authToken;
  final UserSchema user;
  final List<AreaSchema> areas;

  LoginResponseSchema(this.authToken, this.user, this.areas);

  factory LoginResponseSchema.fromJson(dynamic data) {
    return LoginResponseSchema(
      data['auth_token'] as String,
      UserSchema.fromJson(data['user']),
      (data['areas'] as List<dynamic>)
          .map((area) => AreaSchema.fromJson(area))
          .toList(),
    );
  }
}

class GroupSchema {
  final String groupId;
  final String coachId;
  final String coachName;
  final String name;
  final String description;
  final String areaId;

  GroupSchema(this.groupId, this.coachId, this.coachName, this.name,
      this.description, this.areaId);

  factory GroupSchema.fromJson(dynamic data) {
    return GroupSchema(
      data['group_id'] as String,
      data['coach_id'] as String,
      data['coach_name'] as String,
      data['name'] as String,
      data['description'] as String,
      data['area_id'] as String,
    );
  }
}

class GroupInfoSchema {
  final String groupId;
  final String coachName;
  final String name;
  final String areaName;

  GroupInfoSchema(this.groupId, this.coachName, this.name, this.areaName);

  factory GroupInfoSchema.fromJson(dynamic data) {
    return GroupInfoSchema(
      data['group_id'] as String,
      data['coach_name'] as String,
      data['name'] as String,
      data['area_name'] as String,
    );
  }
}

class MeetSchema {
  final String meetId;
  final String groupId;
  final int maxMembers;
  final DateTime meetDate;
  final DateTime startTime;
  final DateTime endTime;
  final int duration;
  final String city;
  final String street;
  final List<UserBaseSchema> members;

  MeetSchema(
      this.meetId,
      this.groupId,
      this.maxMembers,
      this.meetDate,
      this.startTime,
      this.endTime,
      this.duration,
      this.city,
      this.street,
      this.members);

  factory MeetSchema.fromJson(dynamic data) {
    return MeetSchema(
      data['meet_id'] as String,
      data['group_id'] as String,
      data['max_members'] as int,
      DateTime.parse(data['meet_date'] as String),
      DateTime.parse(data['start_time'] as String),
      DateTime.parse(data['end_time'] as String),
      data['duration'] as int,
      data['city'] as String,
      data['street'] as String,
      (data['members'] as List<dynamic>)
          .map((member) => UserBaseSchema.fromJson(member))
          .toList(),
    );
  }
}

class MeetInfoSchema {
  final String meetId;
  final String groupId;
  final String groupName;
  final DateTime meetDate;
  final DateTime startTime;
  final DateTime endTime;
  final int duration;
  final String city;
  final String street;
  final bool full;
  final bool registered;

  MeetInfoSchema(
      this.meetId,
      this.groupId,
      this.groupName,
      this.meetDate,
      this.startTime,
      this.endTime,
      this.duration,
      this.city,
      this.street,
      this.full,
      this.registered);

  factory MeetInfoSchema.fromJson(dynamic data) {
    return MeetInfoSchema(
      data['meet_id'] as String,
      data['group_id'] as String,
      data['group_name'] as String,
      DateTime.parse(data['meet_date'] as String),
      DateTime.parse(data['start_time'] as String),
      DateTime.parse(data['end_time'] as String),
      data['duration'] as int,
      data['city'] as String,
      data['street'] as String,
      data['full'] as bool,
      data['registered'] as bool,
    );
  }
}

class GroupViewInfoSchema {
  final GroupSchema group;
  final List<MeetInfoSchema> meets;
  final bool registered;

  GroupViewInfoSchema(this.group, this.meets, this.registered);

  factory GroupViewInfoSchema.fromJson(dynamic data) {
    return GroupViewInfoSchema(
      GroupSchema.fromJson(data['group']),
      (data['meets'] as List<dynamic>)
          .map((meet) => MeetInfoSchema.fromJson(meet))
          .toList(),
      (data['registered'] as bool),
    );
  }
}

class GroupFullSchema {
  final GroupSchema group;
  final List<MeetSchema> meets;
  final List<UserBaseSchema> members;

  GroupFullSchema(this.group, this.meets, this.members);

  factory GroupFullSchema.fromJson(dynamic data) {
    return GroupFullSchema(
      GroupSchema.fromJson(data['group']),
      (data['meets'] as List<dynamic>)
          .map((meet) => MeetSchema.fromJson(meet))
          .toList(),
      (data['members'] as List<dynamic>)
          .map((meet) => UserBaseSchema.fromJson(meet))
          .toList(),
    );
  }
}

class MyGroupsSchema {
  final List<GroupInfoSchema> inGroups;
  final List<GroupSchema> coachGroups;

  MyGroupsSchema(this.inGroups, this.coachGroups);

  factory MyGroupsSchema.fromJson(dynamic data) {
    return MyGroupsSchema(
      (data['in_groups'] as List<dynamic>)
          .map((group) => GroupInfoSchema.fromJson(group))
          .toList(),
      (data['coach_groups'] as List<dynamic>)
          .map((group) => GroupSchema.fromJson(group))
          .toList(),
    );
  }
}

class MyMeetsSchema {
  final List<MeetInfoSchema> meets;

  MyMeetsSchema(this.meets);

  factory MyMeetsSchema.fromJson(dynamic data) {
    return MyMeetsSchema(
      (data['meets'] as List<dynamic>)
          .map((meet) => MeetInfoSchema.fromJson(meet))
          .toList(),
    );
  }
}

class ViewCoachSchema {
  final UserBaseSchema coach;
  final List<FileSchema> certificates;

  ViewCoachSchema(this.coach, this.certificates);

  factory ViewCoachSchema.fromJson(dynamic data) {
    return ViewCoachSchema(
      UserBaseSchema.fromJson(data['coach']),
      (data['certificates'] as List<dynamic>)
          .map((file) => FileSchema.fromJson(file))
          .toList(),
    );
  }
}
